lib/chargebee/nativeRequest.rb

```diff
@@ -0,0 +1,136 @@
+      # decompress the response. Therefore, we need to manually handle decompression
+      # based on the 'Content-Encoding' header in the response.
+      # https://github.com/ruby/ruby/blob/19c1f0233eb5202403c52b196f1d573893eacab7/lib/net/http/generic_request.rb#L82
+      if headers.keys.any? { |k| k.downcase == 'accept-encoding' } && rheaders[:content_encoding] == 'gzip' && rbody && !rbody.empty?
+        rbody = Zlib::GzipReader.new(StringIO.new(rbody)).read
+      end
+
+      if rcode >= 200 && rcode < 300
+        begin
+          resp = JSON.parse(rbody)
+        rescue JSON::ParserError => e
+          raise handle_json_error(rbody, e)
+        end
+        return Util.symbolize_keys(resp), rheaders
+      else
+        raise handle_for_error(rcode, rbody)
+      end
+    end
+
+    def self.handle_json_error(rbody, e)
+      if rbody.include?("503")
+        raise Error.new("Sorry, the server is currently unable to handle the request due to a temporary overload or scheduled maintenance. Please retry after sometime. \n type: internal_temporary_error, \n http_status_code: 503, \n error_code: internal_temporary_error,\n content: #{rbody.inspect}",e)
+      elsif rbody.include?("504")
+        raise Error.new("The server did not receive a timely response from an upstream server, request aborted. If this problem persists, contact us at support@chargebee.com. \n type: gateway_timeout, \n http_status_code: 504, \n error_code: gateway_timeout,\n content:  #{rbody.inspect}",e)
+      else
+        raise Error.new("Sorry, something went wrong when trying to process the request. If this problem persists, contact us at support@chargebee.com. \n type: internal_error, \n http_status_code: 500, \n error_code: internal_error,\n content:  #{rbody.inspect}",e)
+      end
+    end
+
+    def self.handle_for_error(rcode, rbody)
+      return Error.new("No response returned by ChargeBee API. HTTP status code: #{rcode}") if rcode == 204
+      begin
+        error_obj = JSON.parse(rbody)
+        error_obj = Util.symbolize_keys(error_obj)
+      rescue Exception => e
+        raise Error.new("Error response not in JSON format. The http status code is #{rcode} \n #{rbody.inspect}",e)
+      end
+      type = error_obj[:type]
+      case type
+      when "payment"
+        raise PaymentError.new(rcode, error_obj)
+      when "operation_failed"
+        raise OperationFailedError.new(rcode, error_obj)
+      when "invalid_request"
+        raise InvalidRequestError.new(rcode, error_obj)
+      else
+        raise APIError.new(rcode, error_obj)
+      end
+    end
+    # directly copying headers formatting from rest-client to support backward compatability for rest-client
+    def self.beautify_headers(headers)
+      headers.inject({}) do |out, (key, value)|
+        key_sym = key.tr('-', '_').downcase.to_sym
+
+        # Handle Set-Cookie specially since it cannot be joined by comma.
+        if key.downcase == 'set-cookie'
```

Comment:
You can probably re-use the `key_sym`:

```suggestion
        if key_sym == :set_cookie
```